File: 46-Permutations-全排列/46-Permutations-全排列.java
class Solution {
    List<List<Integer>> res;
    LinkedList<Integer> path = new LinkedList<Integer>();
    public List<List<Integer>> permute(int[] nums) {
        res = new ArrayList<>();
        backtrace(nums);
        return res;
    }

    public void backtrace(int[] nums){
        if(path.size() == nums.length){
            res.add(new ArrayList<Integer>(){{addAll(path);}});
            return;
        }

        for(int i = 0; i < nums.length; i++){
            if(path.contains(nums[i])) continue;
            path.add(nums[i]);
            backtrace(nums);
            path.removeLast();
        }
    }
}

File: 46-Permutations-全排列/46-Permutations-全排列.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 let res
 let path = []
 var permute = function(nums) {
     res = []
     backtrace(nums)
     return res
 };
 
 var backtrace = function(nums){
     if(path.length === nums.length){
         res.push(Array.from(path))
         return
     }
 
     for(let i = 0; i < nums.length; i++){
         if(path.includes(nums[i])) continue
         path.push(nums[i])
         backtrace(nums, i + 1)
         path.pop()
     }
 }